File: ROS/ROS_21_Tutorials/catkin_ws/src/learning_topic/src/velocity_publisher.cpp
/**
 * 该例程将发布turtle1/cmd_vel话题，消息类型geometry_msgs::Twist
 */
 
#include <ros/ros.h>
#include <geometry_msgs/Twist.h> // 包含消息类型的头文件

int main(int argc, char **argv)
{
	// ROS节点初始化, 向master注册
	ros::init(argc, argv, "velocity_publisher"); // 第三个参数为node名字

	// 创建节点句柄, 用于管理节点资源
	ros::NodeHandle n;

	// 创建一个Publisher，发布名为/turtle1/cmd_vel的topic，消息类型为geometry_msgs::Twist，队列长度10
	ros::Publisher turtle_vel_pub = n.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10); // 话题名，10为队列长度, 队列满了ROS会抛弃时间戳最旧的数据

	// 设置循环的频率
	ros::Rate loop_rate(10);

	int count = 0;
	while (ros::ok())
	{
	    // 初始化geometry_msgs::Twist类型的消息
		geometry_msgs::Twist vel_msg;
		vel_msg.linear.x = 0.5;
		vel_msg.angular.z = 0.2;

	    // 发布消息
		turtle_vel_pub.publish(vel_msg);
		ROS_INFO("Publish turtle velocity command[%0.2f m/s, %0.2f rad/s]", 
				vel_msg.linear.x, vel_msg.angular.z);

	    // 按照循环频率延时
	    loop_rate.sleep();
	}

	return 0;
}

File: ROS/ROS_21_Tutorials/catkin_ws/src/learning_topic/scripts/velocity_publisher.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
from geometry_msgs.msg import Twist


def velocity_publisher():
    # ROS节点初始化
    rospy.init_node("velocity_publisher", anonymous=True)

    # 创建一个Publisher， topic名为/turtle1/cmd_vel， message类型为Twist， 队列长度为10
    turtle_vel_publisher = rospy.Publisher("/turtle1/cmd_vel", Twist, queue_size=10)

    # 循环频率为10
    loop_rate = rospy.Rate(10)

    while not rospy.is_shutdown():
        # 创建消息对象
        vel_msg = Twist()
        vel_msg.linear.x = 0.5
        vel_msg.angular.z = 0.5

        # 发布消息
        turtle_vel_publisher.publish(vel_msg)
        rospy.loginfo("Published turtle velocity command: [%.2f m/s, %.2f rad/s]", vel_msg.linear.x, vel_msg.angular.z)

        # 延迟一段时间
        loop_rate.sleep()


if "__main__" == __name__:
    try:
        velocity_publisher()
    except rospy.ROSInterruptException:
        pass
